release-notes/10.0/preview/preview3/runtime.md

```diff
@@ -8,6 +8,94 @@ Here's a summary of what's new in the .NET Runtime in this preview release:
+
+```csharp
+static void Print()
+{
+    string[] words = {"Hello", "World!"};
+    foreach (var str in words)
+    {
+        Console.WriteLine(str);
+    }
+}
+```
+
+The lifetime of `words` is scoped to the `Print` method, and the JIT can avoid heap-allocating the strings `"Hello"` and `"world!"` already. However, the fact that `words` is an array of `strings`, a reference type, would previously stop the JIT from stack-allocating it. Now, the JIT can eliminate every heap allocation in the above example. At the assembly level, the code for `Print` used to look like this:
+```asm
+Program:Print() (FullOpts):
+       push     rbp
+       push     r15
+       push     rbx
+       lea      rbp, [rsp+0x10]
+       mov      rdi, 0x7624BAEF8360      ; System.String[]
+       mov      esi, 2
+       call     CORINFO_HELP_NEWARR_1_OBJ
+       mov      rdi, 0x762534A02D10      ; 'Hello'
+       mov      gword ptr [rax+0x10], rdi
+       mov      rdi, 0x762534A02D30      ; 'World!'
+       mov      gword ptr [rax+0x18], rdi
+       lea      rbx, bword ptr [rax+0x10]
+       mov      r15d, 2
+G_M2084_IG03:  ;; offset=0x0043
+       mov      rdi, gword ptr [rbx]
+       call     [System.Console:WriteLine(System.String)]
+       add      rbx, 8
+       dec      r15d
+       jne      SHORT G_M2084_IG03
+       pop      rbx
+       pop      r15
+       pop      rbp
+       ret
+```
+
+Notice how we call `CORINFO_HELP_NEWARR_1_OBJ` to allocate `words` on the heap. Now, the assembly looks like this:
+```
+Program:Print() (FullOpts):
+       push     rbp
+       push     r15
+       push     rbx
+       sub      rsp, 32
+       lea      rbp, [rsp+0x30]
+       vxorps   xmm8, xmm8, xmm8
+       vmovdqu  ymmword ptr [rbp-0x30], ymm8
+       mov      rdi, 0x700BFC98B0C0      ; System.String[]
+       mov      qword ptr [rbp-0x30], rdi
+       lea      rdi, [rbp-0x30]
+       mov      dword ptr [rdi+0x08], 2
+       lea      rbx, [rbp-0x30]
+       mov      rdi, 0x700C76402D10      ; 'Hello'
+       mov      gword ptr [rbx+0x10], rdi
+       mov      rdi, 0x700C76402D30      ; 'World!'
+       mov      gword ptr [rbx+0x18], rdi
+       add      rbx, 16
+       mov      r15d, 2
+G_M2084_IG03:  ;; offset=0x005A
+       mov      rdi, gword ptr [rbx]
+       call     [System.Console:WriteLine(System.String)]
+       add      rbx, 8
+       dec      r15d
+       jne      SHORT G_M2084_IG03
+       add      rsp, 32
+       pop      rbx
+       pop      r15
+       pop      rbp
+       ret
+```
+
+For more information on stack allocation improvements in the JIT compiler, check out [dotnet/runtime #108913](https://github.com/dotnet/runtime/issues/108913).
+
+## Improved Code Layout
```

Comment:
@AndyAyersMS I'm not sure how we want to highlight this work externally. Right now, it reads pretty bookish, though I'm not sure if there's a way around that.

Reply:
If you want to keep this at a higher level we could create a benchmark with similar pattern and use the `[MemoryDiagnoser]` to show there is now less allocation.

Though it might also run a bit slower since we're now paying to zero init the array in the method rather than getting zeroed memory "for free" from the heap. Structuring the benchmark so it properly reflects when this cost is paid on .NET 9 might be tricky.

Reply:
Sorry, I meant the 3-opt writeup.

Reply:
I think the 3-opt writeup is ok.